### ISSUE TITLE
Integration with Overleaf

### ISSUE THREAD

**Issue by rodluger** [rodluger/showyourwork#22]
This is a popular request, so we need to brainstorm how we might go about integrating ``showyourwork`` with Overleaf. Suggestions welcome!

**Comment by MilesCranmer** [comment]
You can access the git URL of any overleaf paper by going to Menu -> Git. They also have a sync option to a GitHub repository, which they automatically sync changes with.

So, perhaps one could add the git URL as a git submodule, and then create a GitHub action to sync it every commit with the overleaf? Although it could become tricky to deal with rebasing... unless you guarantee that the `showyourwork` repo is in charge of all edits to figures and data, while the overleaf is in charge of LaTeX edits.

**Comment by MilesCranmer** [comment]
> The big difference is that with showyourwork you only commit scripts, not figure files.

I think if it is in a git submodule, the figures would not be stored in the `showyourwork` repository, right? They submodule should just "point" to the particular commit on overleaf, and push the static figures there. Although, it is still true that the overleaf would always need to store the output figure file.

Another point is I know that you can edit the `.latexmkrc` (settings file for `latexmk` command, which is used by overleaf) to actually have overleaf do some preprocessing. For example, the following code:
```sh
$pdflatex = 'bash -c "echo %S && latexdiff old.tex main.tex > .diff.tex && mv .diff.tex diff.tex && pdflatex %O diff.tex"';
```
when stored in ".latexmkrc" in an overleaf repository, will compute the `latexdiff` between old.tex and main.tex, and then render them as the main output PDF. So you can technically have overleaf run an arbitrary bash command when rendering LaTeX, although probably not at the level of generating figures... but who knows.

> But I could imagine the GitHub Action pushing (or force-pushing?) those figure files to the Overleaf repo whenever it runs. 

I think you would want to have the GH action do `git pull --rebase` on the git submodule (so it doesn't overwrite changes to the LaTeX), and then `git push` with the figure changes. 

> Does the Overleaf GitHub repo get updated automatically? Or does the user need to explicitly commit?

Yeah, every single Overleaf paper is actually internally stored as a git repo on their website! So it should be okay to integrate, just by grabbing the correct `git.overleaf.com/...` URL (find via the overleaf menu for any paper) and setting it as a git submodule in the `showyourwork` repo.

**Comment by rodluger** [comment]
I finally had time to sit down and give the integration with Overleaf some thought. I think @MilesCranmer's submodule idea can work, but it will require some major restructuring of `showyourwork` that will break all backward compatibility. Nevertheless, it's a good opportunity to streamline a lot of stuff in the workflow, so I think it's worth it. We'll take steps to make the transition fairly straightforward for existing users.

The basic idea is to use the "Overleaf Git-Bridge" (as opposed to the "Overleaf GitHub Synchronization") functionality in Overleaf (see [here](https://www.overleaf.com/learn/how-to/Using_Git_and_GitHub#The_Overleaf_Git-Bridge) for details). This way, we can make the directory containing the TeX manuscript and the figure files a submodule whose remote is the git repo for the Overleaf project (per @MilesCranmer's recommendation). Builds on GitHub Actions CI will automatically push changes to this remote, and users can similarly sync things locally.

The one big downside of this new workflow is that users will have THREE versions of their article PDF: one locally, one on GitHub (generated from the CI build), and one on Overleaf. I'm not crazy about that, so I'm still thinking through how to best manage synchronization across these three versions. Suggestions welcome.

Below is a list of specific to-do items for version `v0.3.0`, which will implement all of this:

- Add a second GitHub workflow `yml` file to the `showyourwork-template` repository called `setup` and add a `workflow_dispatch` trigger. This will handle initialization of the repository (rather than having that happen automatically upon instantiation). This entails an extra button click on the user side, but the big advantage is that this will allow us to prompt the user for some variables to customize the setup (the `workflow_dispatch` accepts a list of user `input` keys). We'll use this to link the repo to the Overleaf project (see below), and to ask the user about some standard stuff (like their name, email, and ORCID, all optional). Note that if GitHub ever adds support for [template repository variables](https://github.com/github/feedback/discussions/5336), we can streamline this even further. For now, we'll just have to add detailed instructions to the README on how to proceed. We can link to the Actions page for the `setup` workflow, so this really only requires the user to click on that link and then press the "Run workflow" button.

- Get rid of the step that generates `.showyourwork-ms.tex`. Currently, the manuscript is copied to this temporary file, to which the statement `\usepackage{showyourwork}` is added; this is the file that actually gets compiled. Instead, let's just require the tex manuscript to explicitly import the `showyourwork` package, and directly compile the original manuscript. This makes the build _way_ more portable, including for arxiv submissions (change that bit of the workflow, too).

- Add a file `src/figures/showyourwork.py` that exports common paths for the figure scripts (`DATA`, `FIGURES`, `TEX`). This will be useful for the user, since we are changing the directory where figure files should be output to (see below).

- Change the color of figure icons for locally-generated figures. Figures generated on CI should be the gold standard, since they ran in a controlled environment from scratch. This will help especially with the Overleaf stuff, since it make it easy to tell if the article PDF on Overleaf is based on locally-generated or CI-generated figures (see below).

- Add an `overleaf` section to the main config file:

    ```yaml 
    overleaf: 
        id: 5b98bfbf54474d00010421b1 
    ```

    The `id` tells us about the git url for the overleaf repo (https://git.overleaf.com/5b98bfbf54474d00010421b1). It will be populated automatically by the `setup` workflow (see above). This `id` should not be changed by the user (we can think about adding checks and errors if it does change). The `overleaf` section will also accept other keys customizing the integration with Overleaf (details TBD).

- Change the directory structure to

    ```
    /
      ...
      src/
          figure-scripts/
            ...
          data/
            ...
          static-figures/
            ...
          tex/
            figures/
              ...
            ms.tex
            bib.bib
    ``` 

    i.e., move all the TeX stuff into a directory called `tex` nested under `src`, and create a second `figures` directory within `src/tex`, which will now contain the figure files output by the scripts in `src/figures`. I'm also proposing we change the names of the original `figures` and `static` repositories to `figure-scripts` and `static-figures` for clarity.

- Tweak the workflow to copy over all static figures to `src/tex/figures` when building the article. We'll have to change the way we identify static figures inside `showyourwork` (currently we rely on the figure path in the `includegraphics` statement being the `static` directory). TBD.

- Here's where things get fun. If the user did not choose Overleaf integration during `setup` (see above), we're good, and things should mostly work the same way as before (we'll need to add some code to do the customization, like update the `ms.tex` file with the user's name, etc., but that's all straightforward). If, on the other hand, the user specified an associated Overleaf project, we will make the `src/tex` directory a **submodule** pointing to the Overleaf git repo during the `setup` step: 

    ```bash
    git submodule add https://git.overleaf.com/<id> src/tex 
    git submodule update --init --recursive 
    ```

    Note that the Overleaf git repo will be the sole remote for this submodule. We then commit and push this to the GitHub remote.

- Add `src/tex/figures/*` to the top-level `.gitignore` file in the repo template so we don't track them by default. Note that if the `src/tex` file is a submodule, this should have no effect! We actually _want_ to commit the figures within the Overleaf submodule, because that's how Overleaf works (for better or for worse)!

- Code up a command to synchronize changes with Overleaf. This will be run on GitHub Actions as part of the `showyourwork.yml` workflow (add a check to ensure we're on CI). We'll run it in a separate **job** in the workflow (that depends on the `showyourwork` job) so it can pass/fail independently of the main article build. Because the GitHub Actions build always clones a fresh version of the repo, the submodule is up to date with the Overleaf remote by default, so there should not be any merge conflicts (but see below).

    1. Commit changes in the `src/tex` submodule (e.g., updated figures and an updated `showyourwork.sty` file; the latter is ususally git-ignored, but not if the `src/tex` directory is an Overleaf submodule). 
    2. Pull from the `master` branch of the submodule remote in case the Overleaf document was edited during the build process. If there's a merge conflict, fail the job (see below). 
    3. Push any changes to the submodule remote. If they are rejected, fail the job (see below). 

    Merge conflicts are tricky. If the user is following best practices, the only things changing on the Overleaf side are the tex files, and the only things changing on the GitHub side are the figures (and some showyourwork aux files). Even if the user edits a figure file in Overleaf (i.e., deletes one or uploads a new one -- both of which are discouraged!), things should still work: at the beginning of the CI build we clone the Overleaf submodule (up to date with these changes), so all our changes to the figures are just downstream commits from there. A conflict _should_ only happen if the user makes changes to the figures _during_ the CI build process (in between cloning the submodule and pushing the changes). If that does happen, we can just fail the build with an informative error message. Running the CI build again _should_ fix the issue. I may be overlooking other failure modes here, so these details still need to be hashed out.

- Create a Makefile rule to do the same thing (sync w/ Overleaf) locally. Ideally, we'd like the user to primarily sync things with Overleaf via the CI builds, since the figures generated on CI are the truly "reproducible" ones. But I think it's ok as long as we implement the different icon colors for local/CI builds (see above).

- Edit the `make update` rule to provide detailed instructions on what users must do to bring existing repositories up to date with this new version. We could even add linting checks that revert the update if the repository structure doesn't comply with the new standard (TBD). But if users want to enable Overleaf integration, we will require them to start from a fresh template instance.

**Comment by MilesCranmer** [comment]
This looks good to me!

Two quick initial comments:

1. The ID variable for the Overleaf Git repository is the same one as the URL of the paper, so the user can grab it from: `overleaf.com/project/{ID}` rather than finding the git.overleaf URL.
2. Re: synchronization, this is definitely tricky. IIRC, the submodule (in the master branch) is actually fixed to a specific commit! When you `git pull` the repo, the updated submodule is fetched, but it is not actually part of the master branch.  Therefore every time you want to synchronize the submodule, you have to enter this as a new commit to the `showyourwork` repository. One possibility would be to have a scheduled GitHub action: https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#scheduled-events that runs frequently to check whether the Overleaf has been updated (just `git pull`), and if so, update the github with the new edits?

**Comment by MilesCranmer** [comment]
3. Regarding this point:

> Add src/tex/figures/* to the top-level .gitignore file in the repo template so we don't track them by default. Note that if the src/tex file is a submodule, this should have no effect! We actually want to commit the figures within the Overleaf submodule, because that's how Overleaf works (for better or for worse)!

I realized there might be a way around this.

Option 1)

On Overleaf, if you go to "upload file", there is an option for "From External URL". This lets you include an arbitrary file in the Overleaf project defined by some URL - that file will be downloaded and included in builds. However, there are a couple problems with this approach:

- (I think) the URL needs to be publicly accessible.
- I just tested this - it seems like you still need to hit "Refresh" on the image on Overleaf before the updated version shows up. (Maybe there is a way to automatically refresh images. Is there a way to automatically refresh Zotero bib files?)

Option 2)

You can define an arbitrary `.latexmkrc` on any Overleaf document. Within this `.latexmkrc`, you can actually run arbitrary shell commands! Therefore, you could potentially have this latex build file download updates to figures before every build on Overleaf, rather than having this be done on the GitHub side. Thus, the Overleaf would automatically synchronize itself to new figures.

For example, if you write the following:
```
$pdflatex = 'bash -c "custom_command && pdflatex %O main.tex"';
```
into a file named `.latexmkrc` in any Overleaf repo, it will execute your `custom_command` before compiling the pdf!

I think you could even have this `custom_command` run `git pull` on the entirety of the `showyourwork` repo into a hidden folder.

**Comment by rodluger** [comment]
@MilesCranmer thanks for all the ideas! I've been rewriting the code from scratch on the `dev` branch and will soon take a shot at implementing this. I'll keep you posted here.

**Comment by rodluger** [comment]
@MilesCranmer I really like the `.latexmkrc` approach but I think Overleaf may disable network access within the VM that compiles the document. I tried running a `wget` before compiling the document and get this in the stderr output:

```
Resolving github.com (github.com)... failed: Temporary failure in name resolution.
wget: unable to resolve host address ‘github.com’
```

Happens for other domains, too. I couldn't find much documentation on using `.latexmkrc` within Overleaf, so I'm just guessing about internet access being restricted.

**Comment by rodluger** [comment]
I also tested the import logic "From External URL" but it doesn't seem scalable enough for this. Users have to manually import one file at a time, and then manually refresh the files one at a time whenever there are changes. Looking into the git submodule idea next.

**Comment by MilesCranmer** [comment]
Interesting. Too bad it didn't work! Thanks for checking this out.

**Comment by rodluger** [comment]
OK, so regarding the submodule idea. Authentication is needed to push to/pull from the Overleaf submodule, so it's not feasible for showyourwork to rely on this, since third-party users simply won't have access to anything in that submodule (the contents of the `tex` directory, as I had suggested above) unless the credentials are made public, which is not an option. 

I therefore think our best bet is to keep everything within the showyourwork repo and simply copy files over to/from the Overleaf repository when the article is built (and if the user has the correct credentials). I need to think this logic through a bit more, but that's what I'm thinking as of right now.

**Comment by MilesCranmer** [comment]
What about if you select the option to "Sync to GitHub", and then connect the git submodule to the GitHub repo, *not* the overleaf directly? Then you don't need to set up overleaf auth, just the github (I think). This assumes that the overleaf would sync both directions with the GitHub repo, which I'm not sure about.

**Comment by rodluger** [comment]
@MilesCranmer I looked into that today. It could work, but I'm leaning against it because it requires the user to have two separate git repos on GitHub, which is a bit of a pain to set up, and which I don't like on principle because one of those repos has to contain a mix of version-controlled input and output files. 

I coded up the workflow I mentioned above, in which files are pulled/pushed between the repos whenever the article is built with `showyourwork`. It's working pretty well. To get this set up, the user specifies something like this in the config file:

```yaml
overleaf:
  id: 62150dd16134ef045f81d1c8
  push: 
    - src/tex/figures
  pull:
    - src/tex/ms.tex
    - src/tex/bib.bib
```

which tells `showyourwork` which files to pull from Overleaf before the build and which files (in this case a folder) to push to Overleaf after the build. Users can then configure whether to run this syncing every time, only locally, or only on CI.

The workflow is pretty simple, but this way the user has fine-grained control over which remote owns which files. We can also keep the git repo clean from committed figure files. And the setup is much simpler, although I have to work out still how best to bring in existing Overleaf repos.
